src/PIL/JpegImagePlugin.py

```diff
@@ -162,13 +161,15 @@ def APP(self, marker):
                 dpi = float(x_resolution[0]) / x_resolution[1]
             except TypeError:
                 dpi = x_resolution
-            if math.isnan(dpi):
-                raise ValueError
+
+            # Check dpi is a number
+            int(dpi)
+
             if resolution_unit == 3:  # cm
                 # 1 dpcm = 2.54 dpi
                 dpi *= 2.54
```

Comment:
1. This will fail
2. `info["dpi"]` will be undefined type

Reply:
Not following, sorry.

```pycon
>>> int(float("nan"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot convert float NaN to integer
>>> int("test")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'test'
```

So both NaN and your string DPI fail with a ValueError.

https://github.com/uploadcare/pillow-simd/blob/2c018f7d31636158507f1d3afb7c5982813ab757/src/PIL/JpegImagePlugin.py#L172-L177

They are then caught, and the default DPI is set.

See https://github.com/radarhere/Pillow/actions/runs/1069391556 for a passing CI run of this.

Reply:
If `dpi` is a string with a number, it will be stored as is in `info["dpi"]` or rejected depending on `resolution_unit` value. It is very inconsistent.

In my opinion, `info["dpi"]` should always be a tuple of two numbers (floats or ints). The only question is should we parse this number if it is stored as a string. Current implementation in master doesn't this. So my fix is just handles the case when `dpi` is a string. What do you think we can improve?

Reply:
I think you're right. I wasn't trying to handle the case where it was a string, just trying to simplify the checks.